File: python/ml/cv.py
"""cv() — cross-validation within split boundaries.

The 8th primitive. split() creates boundaries, cv() creates rotations.

    s = ml.split(data, "survived", seed=42)   # partition
    c = ml.cv(s, folds=5, seed=42)            # resample within dev
    model = ml.fit(c, "survived", seed=42)    # CV fit + refit on dev
    evidence = ml.assess(model, test=s.test)  # final exam (test on split)
"""

from __future__ import annotations

import numpy as np

from ._types import ConfigError, CVResult, DataError, SplitResult


def cv(
    s: SplitResult,
    *,
    folds: int = 5,
    seed: int,
    stratify: bool = True,
) -> CVResult:
    """Create cross-validation folds within a split's dev partition.

    Takes a SplitResult and creates k-fold rotations within ``.dev``
    (train + valid). The test partition is preserved for ``assess()``.

    This is a second-order primitive: it transforms a partition strategy,
    not raw data. The provenance chain is preserved — the model trained
    on these folds shares the same split_id as the test partition.

    Args:
        s: SplitResult from ``ml.split()``.
        folds: Number of CV folds (default 5).
        seed: Random seed (keyword-only).
        stratify: Stratify folds by target class (default True).
            Only applies to classification tasks.

    Returns:
        CVResult with ``.folds`` (rotations within dev) and ``.k``.
        Test stays on the SplitResult — use ``s.test`` for assess().

    Raises:
        ConfigError: If input is not a SplitResult, folds < 2, or
            no target was set in the original split.
        DataError: If dev partition is too small for requested folds.

    Example:
        >>> import ml
        >>> data = ml.dataset("titanic")
        >>> s = ml.split(data, "survived", seed=42)
        >>> c = ml.cv(s, folds=5, seed=42)
        >>> model = ml.fit(c, "survived", seed=42)
        >>> model.scores_
        {'accuracy_mean': 0.82, 'accuracy_std': 0.03, ...}
        >>> evidence = ml.assess(model, test=s.test)
    """
    # Validate input type
    if not isinstance(s, SplitResult):
        raise ConfigError(
            f"cv() expects a SplitResult from ml.split(), got {type(s).__name__}. "
            "Usage: s = ml.split(data, target, seed=42); c = ml.cv(s, folds=5, seed=42)"
        )

    # Block random folds on temporal data — use cv_temporal() instead
    if getattr(s, "_temporal", False):
        raise ConfigError(
            "cv() creates random folds — not valid for temporal data (future leakage). "
            "Use ml.cv_temporal() instead:\n"
            "  s = ml.split_temporal(data, target, time='col')\n"
            "  c = ml.cv_temporal(s, folds=5, embargo=0)"
        )

    # Validate seed
    if not isinstance(seed, int) or isinstance(seed, bool):
        raise ConfigError(
            f"seed must be an integer, got {type(seed).__name__}: {seed!r}. "
            "Example: seed=42"
        )

    # Validate folds
    if folds < 2:
        raise ConfigError("folds must be >= 2")

    # Get dev data (train + valid)
    dev = s.dev
    target = s._target if hasattr(s, "_target") else None

    if target is None:
        raise ConfigError(
            "cv() requires a target column. "
            "Split with target: s = ml.split(data, 'target', seed=42)"
        )

    n_rows = len(dev)
    if folds > n_rows:
        raise DataError(
            f"Cannot create {folds} folds from {n_rows} dev rows. "
            f"Use folds={max(2, n_rows // 2)} or fewer."
        )

    # Guard against empty folds from stratified CV: folds must not exceed
    # the minority class count, otherwise round-robin allocation leaves
    # some folds with 0 validation rows.
    if stratify and target in dev.columns:
        from .split import _detect_task
        if _detect_task(dev[target]) == "classification":
            min_class_n = dev[target].value_counts().min()
            if folds > min_class_n:
                raise DataError(
                    f"Cannot create {folds} stratified folds: minority class has "
                    f"only {min_class_n} samples. Use folds={max(2, min_class_n)} "
                    f"or fewer, or set stratify=False."
                )

    # Import fold generators from split module
    from .split import _kfold, _stratified_kfold

    # Detect task for stratification
    do_stratify = stratify and target in dev.columns
    if do_stratify:
        from .split import _detect_task
        detected_task = _detect_task(dev[target])
        do_stratify = detected_task == "classification"

    # Create fold indices
    if do_stratify:
        fold_list = list(_stratified_kfold(dev[target].values, k=folds, seed=seed))
    else:
        fold_list = list(_kfold(len(dev), k=folds, seed=seed))

    # Build CVResult — folds on dev only, no test (test stays on SplitResult)
    result = CVResult(
        _data=dev,
        folds=fold_list,
        k=folds,
        target=target,
    )

    return result


def cv_temporal(
    s: SplitResult,
    *,
    folds: int = 5,
    embargo: int = 0,
    window: str = "expanding",
    window_size: int | None = None,
    groups: str | None = None,
) -> CVResult:
    """Create temporal cross-validation folds within a split's dev partition.

    Expanding or sliding window CV, respecting chronological order.
    No future leakage — each fold's training data precedes its validation data.

    When ``groups=`` is set, no group (patient, store, sensor) appears in
    both training and validation within any fold. Handles panel data —
    entities tracked over time.

    Args:
        s: SplitResult from ``ml.split_temporal()``.
        folds: Number of CV folds (default 5).
        embargo: Number of rows to skip between train and valid (gap to prevent
            temporal leakage from autocorrelation). Default 0.
        window: ``"expanding"`` (all prior data) or ``"sliding"`` (fixed window).
        window_size: Required when ``window="sliding"``. Number of rows in
            each training window.
        groups: Column name for group non-overlap (optional). When set,
            rows from the same group are kept together — prevents leakage
            from repeated measurements (patients, stores, devices).

    Returns:
        CVResult with ``.folds`` and ``._temporal=True``.
        Test stays on the SplitResult — use ``s.test`` for assess().

    Raises:
        ConfigError: If input is not a SplitResult, folds < 2, or
            sliding window requested without window_size.
        DataError: If dev partition is too small for requested folds.

    Example:
        >>> s = ml.split_temporal(data, "price", time="date")
        >>> c = ml.cv_temporal(s, folds=5, embargo=10)
        >>> model = ml.fit(c, "price", seed=42)
        >>> evidence = ml.assess(model, test=s.test)

        Panel data (entities over time):

        >>> s = ml.split_temporal(data, "mortality", time="admission_date")
        >>> c = ml.cv_temporal(s, folds=5, embargo=48, groups="patient_id")
    """
    if not isinstance(s, SplitResult):
        raise ConfigError(
            f"cv_temporal() expects a SplitResult, got {type(s).__name__}. "
            "Usage: s = ml.split_temporal(data, target, time='col'); "
            "c = ml.cv_temporal(s, folds=5)"
        )

    if folds < 2:
        raise ConfigError("folds must be >= 2")

    if window not in ("expanding", "sliding"):
        raise ConfigError(
            f"window='{window}' not valid. Choose from: ['expanding', 'sliding']"
        )

    if window == "sliding" and window_size is None:
        raise ConfigError(
            "window_size= is required when window='sliding'. "
            "Example: cv_temporal(s, folds=5, window='sliding', window_size=100)"
        )

    dev = s.dev
    target = s._target if hasattr(s, "_target") else None

    if target is None:
        raise ConfigError(
            "cv_temporal() requires a target column. "
            "Split with target: s = ml.split_temporal(data, 'target', time='col')"
        )

    # Validate groups column
    if groups is not None and groups not in dev.columns:
        raise ConfigError(
            f"groups='{groups}' not found in dev data. "
            f"Available columns: {list(dev.columns[:10])}"
        )

    if embargo < 0:
        raise ConfigError(
            f"embargo must be >= 0, got {embargo}. "
            "Negative embargo would create temporal leakage."
        )

    n = len(dev)
    if folds >= n:
        raise DataError(
            f"Cannot create {folds} temporal CV folds from {n} dev rows. "
            f"Use folds={min(5, n - 1)} or fewer."
        )

    # Build expanding or sliding window folds
    fold_size = n // (folds + 1)
    if fold_size < 1:
        raise DataError(
            f"Too many folds ({folds}) for {n} dev rows. "
            f"Use folds={max(2, n // 2 - 1)} or fewer."
        )

    fold_list = []
    for i in range(folds):
        if window == "sliding" and window_size is not None:
            train_start = max(0, (i + 1) * fold_size - window_size)
        else:
            train_start = 0

        train_end = (i + 1) * fold_size
        valid_start = train_end + embargo

        if i == folds - 1:
            valid_end = n
        else:
            valid_end = (i + 2) * fold_size

        if valid_start >= valid_end:
            continue  # embargo consumed all validation data

        train_idx = np.arange(train_start, train_end)
        valid_idx = np.arange(valid_start, valid_end)

        # Group non-overlap: remove rows from train whose group appears in valid
        if groups is not None:
            valid_groups = set(dev.iloc[valid_idx][groups].unique())
            train_mask = ~dev.iloc[train_idx][groups].isin(valid_groups)
            train_idx = train_idx[train_mask.values]

        fold_list.append((train_idx, valid_idx))

    if not fold_list:
        raise DataError(
            f"Embargo ({embargo}) is too large — no validation data remains. "
            f"Reduce embargo or use fewer folds."
        )

    result = CVResult(
        _data=dev,
        folds=fold_list,
        k=len(fold_list),
        target=target,
    )
    result._temporal = True

    return result


def cv_group(
    s: SplitResult,
    *,
    folds: int = 5,
    groups: str,
    seed: int,
    stratify: bool = True,
) -> CVResult:
    """Create group-aware cross-validation folds within a split's dev partition.

    No group appears in both train and valid within any fold — prevents
    leakage from repeated measurements (patients, subjects, stores, etc.).

    Args:
        s: SplitResult from ``ml.split_group()``.
        folds: Number of CV folds (default 5).
        groups: Column name containing group identifiers.
        seed: Random seed (keyword-only).
        stratify: Stratify groups by target class (default True).

    Returns:
        CVResult with ``.folds`` and ``.k``.
        Test stays on the SplitResult — use ``s.test`` for assess().

    Raises:
        ConfigError: If input is not a SplitResult, folds < 2, or
            groups column not found.
        DataError: If fewer groups than folds.

    Example:
        >>> s = ml.split_group(data, "outcome", groups="patient_id", seed=42)
        >>> c = ml.cv_group(s, folds=5, groups="patient_id", seed=42)
        >>> model = ml.fit(c, "outcome", seed=42)
        >>> evidence = ml.assess(model, test=s.test)
    """
    if not isinstance(s, SplitResult):
        raise ConfigError(
            f"cv_group() expects a SplitResult, got {type(s).__name__}. "
            "Usage: s = ml.split_group(data, target, groups='col', seed=42); "
            "c = ml.cv_group(s, folds=5, groups='col', seed=42)"
        )

    if not isinstance(seed, int) or isinstance(seed, bool):
        raise ConfigError(
            f"seed must be an integer, got {type(seed).__name__}: {seed!r}."
        )

    if folds < 2:
        raise ConfigError("folds must be >= 2")

    dev = s.dev
    target = s._target if hasattr(s, "_target") else None

    if target is None:
        raise ConfigError(
            "cv_group() requires a target column."
        )

    if groups not in dev.columns:
        raise ConfigError(
            f"groups column '{groups}' not found in dev data. "
            f"Available columns: {list(dev.columns[:10])}"
        )

    group_values = dev[groups].values
    n_groups = len(np.unique(group_values))

    if folds > n_groups:
        raise DataError(
            f"Cannot create {folds} folds from {n_groups} groups. "
            f"Use folds<={n_groups}."
        )

    # Import group kfold from split module
    from .split import _group_kfold

    fold_list = list(_group_kfold(group_values, k=folds, seed=seed))

    result = CVResult(
        _data=dev,
        folds=fold_list,
        k=folds,
        target=target,
    )

    return result

File: r/R/cv.R
# -- Cross-validation primitives -----------------------------------------------
#
# cv() creates k-fold rotations *within* an existing split's dev partition.
# Test stays sealed on the original split for ml_assess().
#
# Grammar: split() creates boundaries, cv() creates rotations.

#' Create k-fold cross-validation from a split
#'
#' Takes an existing `ml_split_result` and creates k-fold rotations within its
#' dev partition (train + valid). The test partition stays sealed on the original
#' split for `ml_assess()`.
#'
#' Two primitives, strict separation of concerns: `ml_split()` creates the
#' three-way boundary, `ml_cv()` creates rotations within that boundary.
#'
#' @param s An `ml_split_result` from `ml_split()`
#' @param target Target column name (string)
#' @param folds Number of folds (default 5)
#' @param seed Random seed for fold assignment
#' @param stratify Logical. Stratify folds by target for classification (default TRUE)
#' @returns An `ml_cv_result` that `ml_fit()` accepts directly.
#'   The original split's `$test` remains available via `s$test` for `ml_assess()`.
#' @export
#' @examples
#' s <- ml_split(iris, "Species", seed = 42)
#' c <- ml_cv(s, "Species", folds = 5, seed = 42)
#' model <- ml_fit(c, "Species", seed = 42)
#' model$scores_
ml_cv <- function(s, target, folds = 5L, seed = NULL, stratify = TRUE) {
  if (!inherits(s, "ml_split_result")) {
    config_error("ml_cv() requires an ml_split_result from ml_split().")
  }
  if (isTRUE(s$.temporal)) {
    config_error(
      "temporal split detected -- use ml_cv_temporal() for time-series CV instead."
    )
  }
  if (is.null(seed)) seed <- sample.int(.Machine$integer.max, 1L)
  folds <- as.integer(folds)
  if (folds < 2L) config_error("folds must be an integer >= 2")

  # Use dev = train + valid

  dev_data <- s$dev

  if (!target %in% names(dev_data)) {
    config_error(paste0(
      "target='", target, "' not found in data. Available columns: ",
      paste(names(dev_data), collapse = ", ")
    ))
  }

  n_dev <- nrow(dev_data)
  if (folds > n_dev) {
    config_error(paste0(
      "Cannot create ", folds, " folds from ", n_dev, " dev rows. ",
      "Use folds=", min(n_dev, 10L), " or fewer."
    ))
  }

  withr::local_seed(seed)

  y <- dev_data[[target]]
  task <- .detect_task(y, "auto")

  if (stratify && task == "classification") {
    fold_ids <- .stratified_kfold(y, folds)
  } else {
    fold_ids <- sample(rep(seq_len(folds), length.out = n_dev))
  }

  fold_list <- lapply(seq_len(folds), function(k) {
    list(
      train = which(fold_ids != k),
      valid = which(fold_ids == k)
    )
  })

  cv_result <- list(
    folds  = fold_list,
    data   = dev_data,
    target = target
  )
  class(cv_result) <- "ml_cv_result"
  cv_result
}

#' Create temporal cross-validation from a split
#'
#' Expanding-window CV for time series. Data must already be sorted
#' chronologically (use `ml_split_temporal()` first).
#'
#' @param s An `ml_split_result` from `ml_split_temporal()`
#' @param target Target column name (string)
#' @param folds Number of folds (default 5)
#' @returns An `ml_cv_result` with expanding-window folds
#' @export
#' @examples
#' @param embargo Integer. Number of rows to skip between train end and valid
#'   start (gap to prevent temporal leakage from autocorrelation). Default 0.
#'   Must be >= 0.
#' @param window `"expanding"` (default, all prior rows as train) or
#'   `"sliding"` (fixed-size training window). When `"sliding"`,
#'   `window_size` must also be supplied.
#' @param window_size Integer. Required when `window = "sliding"`. Number of
#'   rows in each training window.
#' @examples
#' df <- data.frame(date = 1:100, x = rnorm(100), y = sample(0:1, 100, TRUE))
#' s <- ml_split_temporal(df, "y", time = "date")
#' c <- ml_cv_temporal(s, "y", folds = 5)
#' # With embargo to prevent autocorrelation leakage:
#' c2 <- ml_cv_temporal(s, "y", folds = 5, embargo = 5L)
ml_cv_temporal <- function(s, target, folds = 5L, embargo = 0L,
                            window = "expanding", window_size = NULL) {
  if (!inherits(s, "ml_split_result")) {
    config_error("ml_cv_temporal() requires an ml_split_result.")
  }
  folds   <- as.integer(folds)
  embargo <- as.integer(embargo)
  if (folds < 2L) config_error("folds must be an integer >= 2")
  if (embargo < 0L) {
    config_error(paste0(
      "embargo must be >= 0, got ", embargo, ". ",
      "Negative embargo would create temporal leakage."
    ))
  }
  if (!window %in% c("expanding", "sliding")) {
    config_error(paste0(
      "window='", window, "' not valid. Choose from: 'expanding', 'sliding'."
    ))
  }
  if (identical(window, "sliding") && is.null(window_size)) {
    config_error(paste0(
      "window_size= is required when window='sliding'. ",
      "Example: ml_cv_temporal(s, 'y', folds=5, window='sliding', window_size=100)"
    ))
  }

  dev_data <- s$dev
  n_dev    <- nrow(dev_data)

  if (folds >= n_dev) {
    config_error(paste0(
      "Cannot create ", folds, " temporal folds from ", n_dev, " dev rows."
    ))
  }

  chunk_size <- n_dev %/% (folds + 1L)
  if (chunk_size < 1L) {
    config_error(paste0(
      "Too many folds (", folds, ") for ", n_dev, " dev rows. ",
      "Use folds=", max(2L, n_dev %/% 2L - 1L), " or fewer."
    ))
  }

  fold_list <- vector("list", folds)
  n_valid_folds <- 0L

  for (k in seq_len(folds)) {
    if (!is.null(window_size) && identical(window, "sliding")) {
      train_start <- max(1L, chunk_size * k - as.integer(window_size) + 1L)
    } else {
      train_start <- 1L
    }
    train_end   <- chunk_size * k
    valid_start <- train_end + embargo + 1L
    valid_end   <- if (k == folds) n_dev else min(train_end + chunk_size, n_dev)

    if (valid_start > valid_end) next  # embargo consumed all validation data

    fold_list[[n_valid_folds + 1L]] <- list(
      train = .safe_seq(train_start, train_end),
      valid = .safe_seq(valid_start, valid_end)
    )
    n_valid_folds <- n_valid_folds + 1L
  }

  if (n_valid_folds == 0L) {
    config_error(paste0(
      "embargo (", embargo, ") is too large -- no validation data remains. ",
      "Reduce embargo or use fewer folds."
    ))
  }

  fold_list <- fold_list[seq_len(n_valid_folds)]

  cv_result <- list(
    folds  = fold_list,
    data   = dev_data,
    target = target
  )
  class(cv_result) <- "ml_cv_result"
  cv_result
}

#' Create group-aware cross-validation from a split
#'
#' No group appears in both train and validation within any fold.
#' Prevents leakage from repeated measurements (patients, stores, sensors).
#'
#' @param s An `ml_split_result` from `ml_split()` or `ml_split_group()`
#' @param target Target column name (string)
#' @param groups Column name identifying groups
#' @param folds Number of folds (default 5)
#' @param seed Random seed for group assignment
#' @returns An `ml_cv_result` with group-aware folds
#' @export
#' @examples
#' df <- data.frame(pid = rep(1:20, each = 5), x = rnorm(100), y = sample(0:1, 100, TRUE))
#' s <- ml_split(df, "y", seed = 42)
#' c <- ml_cv_group(s, "y", groups = "pid", folds = 5, seed = 42)
ml_cv_group <- function(s, target, groups, folds = 5L, seed = NULL) {
  if (!inherits(s, "ml_split_result")) {
    config_error("ml_cv_group() requires an ml_split_result.")
  }
  if (is.null(seed)) seed <- sample.int(.Machine$integer.max, 1L)
  folds <- as.integer(folds)
  if (folds < 2L) config_error("folds must be an integer >= 2")

  dev_data <- s$dev

  if (!groups %in% names(dev_data)) {
    config_error(paste0(
      "groups='", groups, "' not found in data. Available columns: ",
      paste(names(dev_data), collapse = ", ")
    ))
  }

  grp_col <- dev_data[[groups]]
  if (any(is.na(grp_col))) {
    config_error("groups column contains NA values. Drop or impute before CV.")
  }

  unique_grps <- sort(unique(grp_col))
  n_grps <- length(unique_grps)

  if (folds > n_grps) {
    config_error(paste0(
      "Cannot create ", folds, " folds from ", n_grps, " groups. ",
      "Use folds=", min(n_grps, 10L), " or fewer."
    ))
  }

  withr::local_seed(seed)
  perm_grps <- sample(unique_grps)

  # Round-robin assignment of groups to folds
  grp_fold_ids <- rep(seq_len(folds), length.out = n_grps)
  row_fold_ids <- grp_fold_ids[match(as.character(grp_col), as.character(perm_grps))]

  fold_list <- lapply(seq_len(folds), function(k) {
    list(
      train = which(row_fold_ids != k),
      valid = which(row_fold_ids == k)
    )
  })

  cv_result <- list(
    folds  = fold_list,
    data   = dev_data,
    target = target
  )
  class(cv_result) <- "ml_cv_result"
  cv_result
}

# Block access to train/valid/test/dev — CVResult exposes only folds/data/target
#' @export
`$.ml_cv_result` <- function(x, name) {
  if (name %in% c("train", "valid", "dev")) {
    config_error(paste0(
      "No $", name, " on a CVResult. CVResult exposes $folds and $data only. ",
      "Train/valid splits are inside each fold: cv$folds[[1]]$train"
    ))
  }
  if (identical(name, "test")) {
    config_error(
      "CVResult has no $test partition. Test stays on the SplitResult: s$test"
    )
  }
  .subset2(x, name)
}

#' Print ml_cv_result
#' @param x An ml_cv_result object
#' @param ... Ignored
#' @returns The object \code{x}, invisibly.
#' @export
print.ml_cv_result <- function(x, ...) {
  k <- length(x[["folds"]])
  n <- nrow(x[["data"]])
  cat(sprintf("-- CV [%d folds, %d rows] --\n", k, n))
  cat(sprintf("  target : %s\n", x[["target"]]))
  # Show fold sizes
  for (i in seq_len(min(k, 3L))) {
    f <- x[["folds"]][[i]]
    cat(sprintf("  fold %d : %d train, %d valid\n", i, length(f$train), length(f$valid)))
  }
  if (k > 3L) cat(sprintf("  ... (%d more folds)\n", k - 3L))
  cat("\n")
  invisible(x)
}
